### PR TITLE
dbconn: Modify query text to tag source, other metadata

### PR DESCRIPTION
**Pull request by efritz**
This PR modifies the `extendedDriver` type to do a mutating rewrite of SQL query text to add metadata at the front of the query that is more readable from query insights (or other systems that truncate query text).

We're adding the following metadata to each query:

- The query length
- The number of arguments
- A hash of the query (and those hashes will exist in tracing spans for correlation)
- The first relevant caller (function, file, and line number) inferred from the stacktrace. For now we select the canonical caller as the first frame satisfying:
  - Not in a third-party library,
  - Not in `internal/database/{things that deal with conns, helper packages}`, and
  - Not explicit disallowed via `dbconn.SkipFrameForQuerySource(ctx)`. This allows self-aware methods to mark themselves as a layer that shouldn't "own" the query text, but _their_ caller should. There's a usage in the PR as an example. We should detect where we are tagging queries that don't make much sense (e.g., shim layers).

There are plenty of queries around the codebase that start with `-- source: blah/blah/blah`. This will replace those, and keep them up to date!

**TODO**:
- [x] Need to monitor output to see where else we need to add skip frames
- [x] Remove existing `-- source: blah blah blah` from 
- [x] Need to monitor performance to see if we need things like local caches, caller frame or byte array pools

## Test plan

Waiting for feedback before final testing.

### CODE REVIEW

File: internal/database/dbconn/dynamic_metadata.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	metadataLines := []string{hashPrefix, lengthPrefix}
+
+	callerPrefix, ok := getSourceMetadata(ctx)
+	if ok {
+		metadataLines = append(metadataLines, callerPrefix)
+	} else {
+		metadataLines = append(metadataLines, "-- (could not infer source)")
+	}
+
+	fmt.Printf("> %s\n\n\n\n", strings.Join(metadataLines, "\n"))
```
Comment (efritz) [inline internal/database/dbconn/dynamic_metadata.go:38]: See it in action!

Comment (eseliger) [inline internal/database/dbconn/dynamic_metadata.go:38]:
```suggestion
```

**Review by eseliger — APPROVED** [review]
neat 🐱

**Comment by coury-clark** [comment]
I might be missing something, but by hashing the query we will only be able to search for a known query in some kind of log, right? As-is we are able to read logs for a query and reverse lookup the query in source code, which I know I've done a handful of times. I don't necessarily have a strong opinion either way, but I wonder if there would be value in something statically defined 🤷

**Comment by efritz** [comment]
@coury-clark The static part is the location and function name of the caller. The hash is a dynamic lookup that will also be stored in OTel, so that we can scan over postgres logs / query insights that only have the query text and no structured tagging.

**Comment by jhchabran** [comment]
Haven't forgotten this PR, will follow-up tomorrow morning in the train :)

**Comment by jhchabran** [comment]
Also, checked the traces, on the surface level is difference in between with or without this PR is very low. Still, that's not really a useful observation and we'll need to see how it performs on S2 to get a better idea.

**Comment by efritz** [comment]
> Also, checked the traces, on the surface level is difference in between with or without this PR is very low.

@jhchabran Should I clean this up and merge now and we'll add whatever memory caches we find out we need after a trial run?

**Comment by jhchabran** [comment]
> > Also, checked the traces, on the surface level is difference in between with or without this PR is very low.
> 
> @jhchabran Should I clean this up and merge now and we'll add whatever memory caches we find out we need after a trial run?

Yes, the big comment removal PR is green so we can merge this and give it a try.

**Pull request merged**